src/hierarchies/likelihoods/states/multi_ls_state.h

```diff
@@ -0,0 +1,92 @@
+#ifndef BAYESMIX_HIERARCHIES_LIKELIHOODS_STATES_MULTI_LS_STATE_H_
+#define BAYESMIX_HIERARCHIES_LIKELIHOODS_STATES_MULTI_LS_STATE_H_
+
+#include <stan/math/rev.hpp>
+#include <tuple>
+
+#include "algorithm_state.pb.h"
+#include "src/utils/proto_utils.h"
+
+namespace State {
+
+template <typename T>
+Eigen::Matrix<T, Eigen::Dynamic, 1> multi_ls_to_unconstrained(
+    Eigen::Matrix<T, Eigen::Dynamic, 1> mean_in,
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> prec_in) {
+  Eigen::Matrix<T, Eigen::Dynamic, 1> prec_out =
+      stan::math::cov_matrix_free(prec_in);
+  Eigen::Matrix<T, Eigen::Dynamic, 1> out(mean_in.size() + prec_out.size());
+  out << mean_in, prec_out;
+  return out;
+}
+
+template <typename T>
+std::tuple<Eigen::Matrix<T, Eigen::Dynamic, 1>,
+           Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>
+multi_ls_to_constrained(Eigen::Matrix<T, Eigen::Dynamic, 1> in) {
+  double dim_ = 0.5 * (std::sqrt(8 * in.size() + 9) - 3);
+  double dimf;
+  assert(modf(dim_, &dimf) == 0.0);
+  int dim = int(dimf);
+  Eigen::Matrix<T, Eigen::Dynamic, 1> mean(dim);
+  mean << in.head(dim);
+  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> prec(dim, dim);
+  prec = stan::math::cov_matrix_constrain(in.tail(in.size() - dim), dim);
+  return std::make_tuple(mean, prec);
+}
+
+template <typename T>
+T multi_ls_log_det_jac(
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> prec_constrained) {
+  T out = 0;
+  stan::math::positive_constrain(stan::math::cov_matrix_free(prec_constrained),
```

Comment:
I think this is the wrong function, it should be `stan::math::cov_matrix_constrain` or `stan::math::cov_matrix_free`

Reply:
Actually should be `stan::math::cov_matrix_constrain` evaluated on the unconstrained precision, (or -1 * stan::math::cov_matrix_constrain evaluated on the constrained one)

Reply:
It would be good to have some tests!

Reply:
> I think this is the wrong function, it should be `stan::math::cov_matrix_constrain` or `stan::math::cov_matrix_free`

This is a HUGE error, thanks!!

> Actually should be stan::math::cov_matrix_constrain evaluated on the unconstrained precision, (or -1 * stan::math::cov_matrix_constrain evaluated on the constrained one)

This will avoid the constrained-unconstrained transformation and save time, I will test it.